backend/infrahub/graphql/auth/query_permission_checker/object_permission_checker.py

```diff
@@ -0,0 +1,66 @@
+from infrahub.auth import AccountSession
+from infrahub.core import registry
+from infrahub.core.account import ObjectPermission
+from infrahub.core.branch import Branch
+from infrahub.core.constants import PermissionDecision
+from infrahub.database import InfrahubDatabase
+from infrahub.exceptions import PermissionDeniedError
+from infrahub.graphql import GraphqlParams
+from infrahub.graphql.analyzer import InfrahubGraphQLQueryAnalyzer
+from infrahub.utils import extract_camelcase_words
+
+from .interface import CheckerResolution, GraphQLQueryPermissionCheckerInterface
+
+
+class ObjectPermissionChecker(GraphQLQueryPermissionCheckerInterface):
+    account_session: AccountSession
+
+    async def supports(self, db: InfrahubDatabase, account_session: AccountSession, branch: Branch) -> bool:
+        self.account_session = account_session
+        return self.account_session.authenticated
+
+    async def check(
+        self,
+        db: InfrahubDatabase,
+        analyzed_query: InfrahubGraphQLQueryAnalyzer,
+        query_parameters: GraphqlParams,
+        branch: Branch,
+    ) -> CheckerResolution:
+        kinds = await analyzed_query.get_models_in_use(types=query_parameters.context.types)
+
+        # Identify which operation is performed on which kind
+        kind_action_map: dict[str, str] = {}
+        for operation in analyzed_query.operations:
+            for kind in kinds:
+                if operation.name and operation.name.startswith(kind):
+                    # An empty string after prefix removal means a query to "view"
+                    kind_action_map[kind] = operation.name[len(kind) :] or "view"
+
+        # Infer required permissions from the kind/operation map
+        permissions: list[str] = []
+        for kind, action in kind_action_map.items():
+            extracted_words = extract_camelcase_words(kind)
+            permissions.append(
+                str(
+                    # Create a object permission instance just to get its string representation
+                    ObjectPermission(
+                        id="",
+                        branch=branch.name,
+                        namespace=extracted_words[0],
+                        name="".join(extracted_words[1:]),
+                        action=action.lower(),
+                        decision=PermissionDecision.ALLOW.value,
+                    )
+                )
+            )
+
+        for permission in permissions:
+            has_permission = False
+            for permission_backend in registry.permission_backends:
+                has_permission = await permission_backend.has_permission(
+                    db=db, account_id=self.account_session.account_id, permission=permission, branch=branch
+                )
+            if not has_permission:
+                raise PermissionDeniedError(f"You do not have the following permission: {permission}")
+
+        return CheckerResolution.NEXT_CHECKER
```

Comment:
I think it's a small optimisation as some of these can abort when completed.

This is the wrapper function that runs all of them.

```python
class GraphQLQueryPermissionChecker:
    def __init__(self, sub_checkers: list[GraphQLQueryPermissionCheckerInterface]):
        self.sub_checkers = sub_checkers

    async def check(
        self,
        db: InfrahubDatabase,
        account_session: AccountSession,
        analyzed_query: InfrahubGraphQLQueryAnalyzer,
        query_parameters: GraphqlParams,
        branch: Branch,
    ) -> None:
        for sub_checker in self.sub_checkers:
            if await sub_checker.supports(db=db, account_session=account_session, branch=branch):
                resolution = await sub_checker.check(
                    db=db,
                    account_session=account_session,
                    analyzed_query=analyzed_query,
                    query_parameters=query_parameters,
                    branch=branch,
                )
                if resolution == CheckerResolution.TERMINATE:
                    return
        raise PermissionDeniedError("The current account is not authorized to perform this operation")
```

For example ReadOnlyGraphQLPermissionChecker returns TERMINATE to indicate that it would be the last supported check to run if `.supports(). came back as `True`. I think we can remove that part later as once we remove the older checks when they're no longer required if that's fine?